### PR TITLE
4.0.0 branch

### PR DESCRIPTION
**Pull request by g-k**
WIP

Fixes: #520 
Fixes: #566
Fixes: #477

### CODE REVIEW

File: .github/workflows/test.yml [hunk 1]
```diff
@@ -8,7 +8,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: [3.5, 3.6, 3.7, 3.8, pypy3]
+        python-version: [3.6, 3.7, 3.8, pypy3]
```
Comment (hugovk) [inline .github/workflows/test.yml:11]: 3.9 is also available

```suggestion
        python-version: [3.6, 3.7, 3.8, 3.9, pypy3]
```

File: setup.py [hunk 2]
```diff
@@ -54,10 +54,7 @@ def get_version():
...
         'Operating System :: OS Independent',
         'Programming Language :: Python',
-        'Programming Language :: Python :: 2',
-        'Programming Language :: Python :: 2.7',
-        'Programming Language :: Python :: 3',
-        'Programming Language :: Python :: 3.5',
+        'Programming Language :: Python :: 3 :: Only',
         'Programming Language :: Python :: 3.6',
         'Programming Language :: Python :: 3.7',
         'Programming Language :: Python :: 3.8',
```
Comment (hugovk) [inline setup.py:60]:
```suggestion
        'Programming Language :: Python :: 3.8',
        'Programming Language :: Python :: 3.9',
```

File: tox.ini [hunk 3]
```diff
@@ -2,8 +2,8 @@
 
 [tox]
 envlist =
-    py{27,35,36,37,38,py,py3}
-    py{27,35,36,37,38}-build-no-lang
+    py{36,37,38,py3}
+    py{36,37,38}-build-no-lang
```
Comment (hugovk) [inline tox.ini:6]:
```suggestion
    py{36,37,38,39,py3}
    py{36,37,38,39}-build-no-lang
```

File: tox.ini [hunk 4]
```diff
@@ -61,10 +49,17 @@ basepython = python3.8
 changedir = scripts
 deps =
     -rrequirements-dev.txt
-    black
 commands =
     ./run_tests.sh format-check
 
+[testenv:type-check]
+basepython = python3.8
```
Comment (hugovk) [inline tox.ini:56]:
```suggestion
basepython = python3.9
```

File: .github/workflows/lint.yml [hunk 5]
```diff
@@ -10,7 +10,7 @@ jobs:
       matrix:
         python-version: [3.8]
```
Comment (hugovk) [inline .github/workflows/lint.yml:11]:
```suggestion
        python-version: [3.9]
```

**Review by hugovk — APPROVED** [review]

File: .github/workflows/test.yml (hunk 1 above)
Comment (g-k) [reply .github/workflows/test.yml:11]: Yes, but tests still fail due to the urlparse changes mentioned #536

**Comment by g-k** [comment]
Splitting this up to provide a cleaner migration path to library consumers.

**Pull request closed without merging**